src/profile-logic/call-node-info.js

```diff
@@ -274,51 +270,298 @@ export class CallNodeInfoImpl implements CallNodeInfo {
+// arrays in this struct are fixed-size typed arrays.
+// The number of roots is the same as the number of functions in the funcTable.
+type InvertedRootCallNodeTable = {|
+  category: Int32Array, // IndexIntoFuncTable -> IndexIntoCategoryList
+  subcategory: Int32Array, // IndexIntoFuncTable -> IndexIntoSubcategoryListForCategory
+  innerWindowID: Float64Array, // IndexIntoFuncTable -> InnerWindowID
+  // IndexIntoNativeSymbolTable: all frames that collapsed into this call node inlined into the same native symbol
+  // -1: divergent: some, but not all, frames that collapsed into this call node were inlined, or they are from different symbols
+  // null: no inlining
+  sourceFramesInlinedIntoSymbol: Array<IndexIntoNativeSymbolTable | -1 | null>,
+  // The (exclusive) end of the suffix order index range for each root node.
+  // The beginning of the range is given by suffixOrderIndexRangeEnd[i - 1], or by
+  // zero. This is possible because both the inverted root order and the suffix order
+  // are determined by the func order.
+  suffixOrderIndexRangeEnd: Uint32Array, // IndexIntoFuncTable -> SuffixOrderIndex,
+  length: number,
+|};
+
+// Information about the non-root nodes of the inverted call tree. This table
+// grows on-demand, as new inverted call nodes are materialized.
+type InvertedNonRootCallNodeTable = {|
+  prefix: InvertedCallNodeHandle[],
+  func: IndexIntoFuncTable[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoFuncTable
+  pathHash: string[], // IndexIntoInvertedNonRootCallNodeTable -> string
+  category: IndexIntoCategoryList[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoCategoryList
+  subcategory: IndexIntoSubcategoryListForCategory[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoSubcategoryListForCategory
+  innerWindowID: InnerWindowID[], // IndexIntoInvertedNonRootCallNodeTable -> InnerWindowID
+  // IndexIntoNativeSymbolTable: all frames that collapsed into this call node inlined into the same native symbol
+  // -1: divergent: some, but not all, frames that collapsed into this call node were inlined, or they are from different symbols
+  // null: no inlining
+  sourceFramesInlinedIntoSymbol: Array<IndexIntoNativeSymbolTable | -1 | null>,
+  suffixOrderIndexRangeStart: SuffixOrderIndex[], // IndexIntoInvertedNonRootCallNodeTable -> SuffixOrderIndex
+  suffixOrderIndexRangeEnd: SuffixOrderIndex[], // IndexIntoInvertedNonRootCallNodeTable -> SuffixOrderIndex
+
+  // Non-null for non-root nodes whose children haven't been created yet.
+  // For a non-root node x of the inverted tree, let k = depth[x] its depth in the inverted tree,
+  // and deepNodes = deepNodesForSuffixOrderIndexRange[x] be its non-null deep nodes.
+  // Then, for every index i in suffixOrderIndexRangeStart[x]..suffixOrderIndexRangeEnd[x],
+  // the k'th prefix node of suffixOrderedCallNodes[i] is stored at deepNodes[i - suffixOrderIndexRangeStart[x]].
```

Comment:
> I think I need a concrete example.

Yup, this definitely needs an example. I wasn't sure where to put it but I'll try to find a place.

> Also I find the name `deepNodesForSuffixOrderIndexRange` confusing: I'd remove the `Range` at the end because the index isn't a range, it is non root index handle.

Ah, interesting. Yes, it is a non-root node handle. But that non-root node has a suffix order index range. And for that range, these are the corresponding deep nodes.

Reply:
I've renamed this field to just `deepNodes`.